crates/sui-core/src/par_index_live_object_set.rs

```diff
@@ -0,0 +1,104 @@
+// Copyright (c) Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::authority::authority_store_tables::LiveObject;
+use crate::authority::AuthorityStore;
+use std::time::Instant;
+use sui_types::base_types::ObjectID;
+use sui_types::object::Object;
+use sui_types::storage::error::Error as StorageError;
+use tracing::info;
+
+/// Make `LiveObjectIndexer`s for parallel indexing of the live object set
+pub trait ParMakeLiveObjectIndexer: Sync {
+    type ObjectIndexer: LiveObjectIndexer;
+
+    fn make_live_object_indexer(&self) -> Self::ObjectIndexer;
+}
+
+/// Represents an instance of a indexer that operates on a subset of the live object set
+pub trait LiveObjectIndexer {
+    /// Called on each object in the range of the live object set this indexer task is responsible
+    /// for.
+    fn index_object(&mut self, object: Object) -> Result<(), StorageError>;
+
+    /// Called once the range of objects this indexer task is responsible for have been processed
+    /// by calling `index_object`.
+    fn finish(self) -> Result<(), StorageError>;
+}
+
+/// Utility for iterating over, and indexing, the live object set in parallel
+///
+/// This is done by dividing the addressable ObjectID space into smaller, disjoint sets and
+/// operating on each set in parallel in a separate thread. User's will need to implement the
+/// `ParMakeLiveObjectIndexer` trait which will be used to make N `LiveObjectIndexer`s which will
+/// then process one of the disjoint parts of the live object set.
+pub fn par_index_live_object_set<T: ParMakeLiveObjectIndexer>(
+    authority_store: &AuthorityStore,
+    make_indexer: &T,
+) -> Result<(), StorageError> {
+    info!("Indexing Live Object Set");
+    let start_time = Instant::now();
+    std::thread::scope(|s| -> Result<(), StorageError> {
+        let mut threads = Vec::new();
+        const BITS: u8 = 5;
+        for index in 0u8..(1 << BITS) {
+            threads.push(s.spawn(move || {
+                let object_indexer = make_indexer.make_live_object_indexer();
+                live_object_set_index_task(index, BITS, authority_store, object_indexer)
+            }));
+        }
+
+        // join threads
+        for thread in threads {
+            thread.join().unwrap()?;
+        }
+
+        Ok(())
+    })?;
+
+    info!(
+        "Indexing Live Object Set took {} seconds",
+        start_time.elapsed().as_secs()
+    );
+
+    Ok(())
+}
+
+fn live_object_set_index_task<T: LiveObjectIndexer>(
+    task_id: u8,
+    bits: u8,
+    authority_store: &AuthorityStore,
+    mut object_indexer: T,
+) -> Result<(), StorageError> {
+    let mut id_bytes = [0; ObjectID::LENGTH];
+    id_bytes[0] = task_id << (8 - bits);
+    let start_id = ObjectID::new(id_bytes);
+
+    id_bytes[0] |= (1 << (8 - bits)) - 1;
+    for element in id_bytes.iter_mut().skip(1) {
+        *element = u8::MAX;
+    }
+    let end_id = ObjectID::new(id_bytes);
+
```

Comment:
I'm not going to change this logic now. This was original code written for the formal snapshot restore and works by partitioning the address space out into contiguous chunks so that each task could use a bounded iterator to iterate through a section of contiguous  objects. The way that object ids are calculated, by using a hash function already, means that the objects should already be generally uniformly distributed across the address space. Doing as you suggested could likely will lead to slower performance because then you aren't reading from the db in parallel but are instead reading sequentially and then sending those objects to each task. Also `object.hash() % num_tasks` is essentially what this is doing just in a slightly more efficient manor.